src/c_box.rs

```diff
@@ -0,0 +1,94 @@
+use std::ffi::c_void;
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::ptr::drop_in_place;
+use std::ptr::NonNull;
+
+pub type FnFree = unsafe extern "C" fn(ptr: *const u8, cookie: *mut c_void);
+
+/// A `free` "closure", i.e. a [`FnFree`] and an enclosed context [`Self::cookie`].
+#[derive(Debug)]
+pub struct Free {
+    pub free: FnFree,
+    pub cookie: *mut c_void,
+}
+
+impl Free {
+    pub unsafe fn free(&self, ptr: *mut c_void) {
+        (self.free)(ptr as *const u8, self.cookie)
+    }
+}
+
+/// A C/custom [`Box`].
+///
+/// That is, it is analogous to a [`Box`],
+/// but it lets you set a C-style `free` `fn` for deallocation
+/// instead of the normal [`Box`] (de)allocator.
+/// It can also store a normal [`Box`] as well.
+#[derive(Debug)]
+pub struct CBox<T: ?Sized> {
```

Comment:
I checked the codegen/layout on [godbolt](https://godbolt.org/z/cjjYWEsGq):

```rust
use std::ffi::c_void;
use std::marker::PhantomData;
use std::ops::Deref;
use std::ptr::drop_in_place;
use std::ptr::NonNull;

pub fn box_into_raw<T: ?Sized>(r#box: Box<T>) -> NonNull<T> {
    let raw = Box::into_raw(r#box);
    // Safety: [`Box::into_raw`] never returns null.
    unsafe { NonNull::new_unchecked(raw) }
}

pub type FnFree = unsafe extern "C" fn(ptr: *const u8, cookie: *mut c_void);

/// A `free` "closure", i.e. a [`FnFree`] and an enclosed context [`Self::cookie`].
// #[derive(Debug)]
pub struct Free {
    pub free: FnFree,
    pub cookie: *mut c_void,
}

impl Free {
    pub unsafe fn free(&self, ptr: *mut c_void) {
        (self.free)(ptr as *const u8, self.cookie)
    }
}

/// A C/custom [`Box`].
///
/// That is, it is analogous to a [`Box`],
/// but it lets you set a C-style `free` `fn` for deallocation
/// instead of the normal [`Box`] (de)allocator.
/// It can also store a normal [`Box`] as well.
// #[derive(Debug)]
pub struct CBox<T: ?Sized> {
    data: NonNull<T>,
    /// If [`None`], [`Self::data`] should be freed as a [`Box`].
    free: Option<Free>,
    _phantom: PhantomData<T>,
}

impl<T: ?Sized> AsRef<T> for CBox<T> {
    fn as_ref(&self) -> &T {
        unsafe { self.data.as_ref() }
    }
}

impl<T: ?Sized> Deref for CBox<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.as_ref()
    }
}

impl<T: ?Sized> Drop for CBox<T> {
    fn drop(&mut self) {
        let ptr = self.data.as_ptr();
        match &self.free {
            None => {
                // Safety: If [`Self::free`] is [`None`],
                // then [`Self::data`] is a [`Box`].
                // See [`Self::from_box`].
                let _ = unsafe { Box::from_raw(ptr) };
            }
            Some(free) => {
                // Safety: See below.
                // The [`FnFree`] won't run Rust's `fn drop`,
                // so we have to do this ourselves first.
                unsafe { drop_in_place(ptr) };
                let ptr = ptr.cast();
                // Safety: If [`Self::free`] is [`Some`],
                // then that [`FnFree`] should be used for `free`ing.
                // See [`Self::from_c`].
                unsafe { free.free(ptr) }
            }
        }
    }
}

impl<T: ?Sized> CBox<T> {
    pub fn from_c(data: NonNull<T>, free: Free) -> Self {
        Self {
            data,
            free: Some(free),
            _phantom: PhantomData,
        }
    }

    pub fn from_box(data: Box<T>) -> Self {
        Self {
            data: box_into_raw(data),
            free: None,
            _phantom: PhantomData,
        }
    }
}

pub enum CBoxEnum<T: ?Sized> {
    Rust(Box<T>),
    C {
        data: NonNull<T>,
        free: Free,
        _phantom: PhantomData<T>,
    },
}

impl<T: ?Sized> AsRef<T> for CBoxEnum<T> {
    fn as_ref(&self) -> &T {
        match self {
            Self::Rust(r#box) => r#box.as_ref(),
            Self::C { data, .. } => unsafe { data.as_ref() },
        }
    }
}

impl<T: ?Sized> Deref for CBoxEnum<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.as_ref()
    }
}

impl<T: ?Sized> Drop for CBoxEnum<T> {
    fn drop(&mut self) {
        match self {
            Self::Rust(_) => {}, // Drop normally.
            Self::C { data, free, .. } => {
                let ptr = data.as_ptr();
                unsafe { drop_in_place(ptr) };
                let ptr = ptr.cast();
                unsafe { free.free(ptr) }
            }
        }
    }
}

impl<T: ?Sized> CBoxEnum<T> {
    pub fn from_c(data: NonNull<T>, free: Free) -> Self {
        Self::C {
            data,
            free,
            _phantom: PhantomData,
        }
    }

    pub fn from_box(data: Box<T>) -> Self {
        Self::Rust(data)
    }
}

pub fn f(data: &Box<[u8]>, i: usize) -> u8 {
    data[i]
}

pub fn g(data: &CBox<[u8]>, i: usize) -> u8 {
    data[i]
}

pub fn h(data: &CBoxEnum<[u8]>, i: usize) -> u8 {
    data[i]
}
```

`f`, `g`, and `h` are all identical, so I'll switch to the `enum` version as it definitely is simpler.  I didn't realize the optimizer would be smart enough to ensure data ptr is in the same place for both variants.

Reply:
Done in fc43ddc.